File: maze/maze.go
package maze

import (
	"errors"
	"fmt"
	"slices"
)

// Dimensions defines the actual number of cells that make up the maze.
// NumCols is the number of cells along the horizontal edge (columns).
// NumRows is the number of cells along the vertical edge (rows).
type Dimensions struct {
	NumCols       int
	NumRows       int
	StartPosition [2]int
	FinalPosition [2]int
}

// pathStep stores the DFS stack entry for the current maze path.
// moveDirection is the direction used to enter this cell.
// corridorLength is the number of consecutive moves taken in that same direction.
type pathStep struct {
	cellNo         int
	moveDirection  direction
	corridorLength int
}

// GenerateMaze converts the created grid view playing field into a series on paths and walls.
// The Maze is created such that only a single path can exists between the starting point and
// and the goal.
func (config *Dimensions) GenerateMaze(weight WallWeight) ([][]string, error) {
	totalCells, errValidate := config.validateMazeGenerationInputs(weight)
	if errValidate != nil {
		config.resetPositions()
		return nil, errValidate
	}

	// Maze size controls how strongly generation should resist long straight corridors.
	navigationProfile := GetNavigationProfile(*config)

	// The visited set is scoped to a single generation so repeated runs do not leak traversal state.
	visitedCells := make([]bool, totalCells+1)
	startPos, errStart := config.getStartPosition()
	if errStart != nil {
		config.resetPositions()
		return nil, fmt.Errorf("select maze start position: %w", errStart)
	}

	// DFS starts from a random edge cell and grows a spanning tree from there.
	longestPathLength, finalCell := 1, startPos
	cellsPath := []pathStep{{cellNo: startPos}}
	currentPos, visitedCount := startPos, 1

	maze, err := config.CreatePlayingField(weight)
	if err != nil {
		return nil, err
	}

	startAddr := config.GetCellAddress(startPos)
	config.StartPosition = [2]int{startAddr.MiddleCenter[0], startAddr.MiddleCenter[1]}

	visitedCells[currentPos] = true

	for visitedCount < totalCells {
		var neighbors []int

		// Walk backward until we find the next branch that still has an unvisited exit.
		cellsPath, currentPos, neighbors, err = config.backtrackToBranch(cellsPath, visitedCells)
		if err != nil {
			config.resetPositions()
			return nil, err
		}

		// Corridor shaping happens here; the returned cell still preserves DFS behavior.
		nextChoice, nextChoiceErr := config.chooseNextCell(
			neighbors, cellsPath[len(cellsPath)-1], navigationProfile, visitedCells,
		)
		if nextChoiceErr != nil {
			config.resetPositions()
			return nil, fmt.Errorf("select next maze cell: %w", nextChoiceErr)
		}

		if visitedCells[nextChoice.cellNo] {
			continue
		}

		visitedCells[nextChoice.cellNo] = true
		visitedCount++

		config.createPath(maze, currentPos, nextChoice.cellNo)
		cellsPath = append(cellsPath, nextChoice)

		// The deepest point reached from the chosen start becomes the player's target cell.
		if len(cellsPath) > longestPathLength {
			finalCell, longestPathLength = nextChoice.cellNo, len(cellsPath)
		}
	}

	finalAddr := config.GetCellAddress(finalCell)
	config.FinalPosition = [2]int{finalAddr.MiddleCenter[0], finalAddr.MiddleCenter[1]}

	if positionErr := config.validateGeneratedPositions(); positionErr != nil {
		config.resetPositions()
		return nil, positionErr
	}

	config.optimizeMaze(weight, maze)
	return maze, nil
}

// chooseNextCell selects the next unvisited neighbor while applying the configured corridor
// length cap and least-neighbors bias. The returned choice preserves the DFS spanning-tree
// behavior.
func (config *Dimensions) chooseNextCell(
	neighbors []int, currentState pathStep, profile NavigationProfile, visitedCells []bool,
) (pathStep, error) {
	var (
		allCount    int
		lengthCount int

		// These fixed-size arrays avoid per-step heap growth while we score up to four neighbors.
		allChoices        [mazeEdgeNeighborCount]pathStep
		withinLengthLimit [mazeEdgeNeighborCount]pathStep
	)

	for _, neighbor := range neighbors {
		straightLength := 1
		nextDirection := config.directionBetween(currentState.cellNo, neighbor)

		if nextDirection == currentState.moveDirection {
			straightLength += currentState.corridorLength
		}

		choice := pathStep{
			cellNo:         neighbor,
			moveDirection:  nextDirection,
			corridorLength: straightLength,
		}
		allChoices[allCount] = choice
		allCount++

		// Caps how long a straight run can go before being forced to bend.
		if straightLength <= profile.MaxCorridorLength {
			withinLengthLimit[lengthCount] = choice
			lengthCount++
		}
	}

	// Start with every valid neighbor, then narrow down only if the profile says this corridor is too long.
	choices := allChoices[:allCount]
	if lengthCount > 0 {
		choices = withinLengthLimit[:lengthCount]
	}

	if len(choices) > 1 {
		biasRoll, err := secureRandomIndex(percentScale)
		if err != nil {
			return pathStep{}, err
		}

		if biasRoll < profile.LeastNeighborsBias {
			// Prefer the candidate with the fewest remaining unvisited neighbors of its own. A
			// low-neighbor-count cell gets "used up" cleanly by visiting it now, leaving nothing
			// behind for some later, unrelated branch to claim and retroactively turn this cell
			// into a junction. This is the mechanism that actually controls branching — unlike
			// corridor length, neighbor count directly predicts whether a cell will be orphaned.
			var (
				leastPopulatedCount int
				fewestRemaining     = mazeEdgeNeighborCount + 1
				leastPopulated      [mazeEdgeNeighborCount]pathStep
			)

			for _, choice := range choices {
				remaining := len(config.getPresentNeighbors(choice.cellNo, visitedCells))
				switch {
				case remaining < fewestRemaining:
					fewestRemaining = remaining
					leastPopulated[0] = choice
					leastPopulatedCount = 1
				case remaining == fewestRemaining:
					leastPopulated[leastPopulatedCount] = choice
					leastPopulatedCount++
				}
			}
			choices = leastPopulated[:leastPopulatedCount]
		}
	}

	// A final random pick keeps mazes of the same size from following the same local path every run.
	nextChoiceIndex, err := secureRandomIndex(len(choices))
	if err != nil {
		return pathStep{}, err
	}

	return choices[nextChoiceIndex], nil
}

// validateMazeGenerationInputs checks the generation inputs before any traversal state is allocated.
func (config *Dimensions) validateMazeGenerationInputs(weight WallWeight) (int, error) {
	if !weight.IsValid() {
		return 0, fmt.Errorf(
			"invalid wall weight: %s. allowed values are %s, %s, and %s",
			weight, WallWeightRegular, WallWeightMedium, WallWeightBold,
		)
	}

	if config.NumCols <= 0 || config.NumRows <= 0 {
		return 0, fmt.Errorf(
			"invalid maze dimensions: numCols=%d numRows=%d. both values must be greater than zero",
			config.NumCols, config.NumRows,
		)
	}

	totalCells := config.NumCols * config.NumRows
	if totalCells < minPlayableMazeCells {
		return 0, fmt.Errorf(
			"invalid maze dimensions: numCols=%d numRows=%d. at least two cells are required to create distinct endpoints",
			config.NumCols,
			config.NumRows,
		)
	}

	return totalCells, nil
}

// resetPositions clears the generated maze endpoints before returning a generation error.
func (config *Dimensions) resetPositions() {
	config.StartPosition = [2]int{}
	config.FinalPosition = [2]int{}
}

// validateGeneratedPositions ensures a generated maze exposes two distinct endpoints for gameplay.
func (config *Dimensions) validateGeneratedPositions() error {
	if slices.Equal(config.StartPosition[:], config.FinalPosition[:]) {
		return fmt.Errorf(
			"generate maze: start and final positions must differ (numCols=%d numRows=%d)",
			config.NumCols, config.NumRows,
		)
	}
	return nil
}

// createPath creates a path on the common wall between the current and the new cell.
// A path is created by replacing the wall characters with the respective number of blank spaces.
// Wall characters are defined by the wall weight used while creating the grid view.
func (config *Dimensions) createPath(maze [][]string, currentCellNo, newCellNo int) {
	addr := config.GetCellAddress(currentCellNo)
	neighbors := config.GetCellNeighbors(currentCellNo)

	switch newCellNo {
	case neighbors.Bottom:
		maze[addr.BottomCenter[0]][addr.BottomCenter[1]] = "   "

	case neighbors.Left:
		maze[addr.MiddleLeft[0]][addr.MiddleLeft[1]] = " "

	case neighbors.Right:
		maze[addr.MiddleRight[0]][addr.MiddleRight[1]] = " "

	case neighbors.Top:
		maze[addr.TopCenter[0]][addr.TopCenter[1]] = "   "
	}
}

// getPresentNeighbors returns a slice of the neigboring cells associated with the cell number provided.
// Only neighboring cells with no common paths to others cells that are returned. i.e. Non-Visited Cells.
func (config *Dimensions) getPresentNeighbors(cellNo int, visitedCells []bool) []int {
	neighbors := config.GetCellNeighbors(cellNo)
	presentCells := make([]int, 0, mazeEdgeNeighborCount)

	// Only unvisited neighbors are eligible so the maze remains a spanning tree with one unique route between cells.
	if neighbors.Bottom != 0 && !visitedCells[neighbors.Bottom] {
		presentCells = append(presentCells, neighbors.Bottom)
	}

	if neighbors.Left != 0 && !visitedCells[neighbors.Left] {
		presentCells = append(presentCells, neighbors.Left)
	}

	if neighbors.Right != 0 && !visitedCells[neighbors.Right] {
		presentCells = append(presentCells, neighbors.Right)
	}

	if neighbors.Top != 0 && !visitedCells[neighbors.Top] {
		presentCells = append(presentCells, neighbors.Top)
	}

	return presentCells
}

// backtrackToBranch rewinds the DFS stack until it finds a cell with at least one unvisited neighbor.
func (config *Dimensions) backtrackToBranch(cellsPath []pathStep, visitedCells []bool) ([]pathStep, int, []int, error) {
	for len(cellsPath) > 0 {
		currentPos := cellsPath[len(cellsPath)-1].cellNo
		neighbors := config.getPresentNeighbors(currentPos, visitedCells)
		if len(neighbors) > 0 {
			return cellsPath, currentPos, neighbors, nil
		}

		// Dead ends trigger backtracking through the current DFS path until a branch point is found.
		cellsPath = cellsPath[:len(cellsPath)-1]
	}

	return nil, 0, nil, errors.New("generate maze: no unvisited neighbors available during traversal")
}

// directionBetween identifies how traversal would move from the current cell into the chosen neighbor.
func (config *Dimensions) directionBetween(currentCell, nextCell int) direction {
	neighbors := config.GetCellNeighbors(currentCell)

	switch nextCell {
	case neighbors.Top:
		return directionUp
	case neighbors.Bottom:
		return directionDown
	case neighbors.Left:
		return directionLeft
	case neighbors.Right:
		return directionRight
	default:
		return directionNone
	}
}

// getStartPosition returns the cell which becomes the maze traversal starting position.
// The starting position can only be a cell along the  maze edges i.e. has less than four
// neighbors. When getStartPosition is called, all cells are have no common paths to other cells.
func (config *Dimensions) getStartPosition() (int, error) {
	totalCells := config.NumCols * config.NumRows
	for {
		randCellNo, err := secureRandomIndex(totalCells)
		if err != nil {
			return 0, err
		}

		randCellNo++

		// Edge cells have fewer than four neighbors, which guarantees the player starts on the maze boundary.
		if countNeighbors(config.GetCellNeighbors(randCellNo)) < mazeEdgeNeighborCount {
			return randCellNo, nil
		}
	}
}

// optimizeMaze replaces some wall characters so as the maze can
// be more clear and sharp when printed on the terminal.
func (config *Dimensions) optimizeMaze(weight WallWeight, maze [][]string) {
	chars, err := getWallCharacters(weight)
	if err != nil {
		panic(err)
	}

	// Corner cleanup swaps in horizontal glyphs where two open passages meet for a cleaner terminal render.
	for cell := 1; cell <= (config.NumCols * config.NumRows); cell++ {
		addr := config.GetCellAddress(cell)

		config.replaceChar(addr.BottomRight, chars[2], maze)
		config.replaceChar(addr.TopRight, chars[2], maze)
	}
}

// replaceChar switches left and right wall character with a top and bottom wall character.
func (config *Dimensions) replaceChar(point [2]int, replChar string, maze [][]string) {
	elemTop, elemBottom := "", ""
	lenTop, lenBottom := false, false

	// checks if the top point in relation to the given point can be calculated
	if (point[0] - 1) > 0 {
		elemTop = maze[point[0]-1][point[1]]
		lenTop = true
	}

	// checks if the bottom point in relation to the given point can be calculated
	if (point[0] + 1) <= (config.NumRows * cellSpan) {
		elemBottom = maze[point[0]+1][point[1]]
		lenBottom = true
	}

	x, y := point[0], point[1]
	switch {
	case !lenTop && lenBottom && isSpaceFound(elemBottom):
		maze[x][y] = replChar

	case lenTop && !lenBottom && isSpaceFound(elemTop):
		maze[x][y] = replChar

	case lenTop && lenBottom && isSpaceFound(elemBottom) && isSpaceFound(elemTop):
		maze[x][y] = replChar
	}
}

func countNeighbors(neighbors CellNeighbors) int {
	count := 0

	if neighbors.Bottom != 0 {
		count++
	}

	if neighbors.Left != 0 {
		count++
	}

	if neighbors.Right != 0 {
		count++
	}

	if neighbors.Top != 0 {
		count++
	}

	return count
}

File: frontend/app/maze.ts
import { CONFIG } from "./config"
import { createMazeDimensions, isSpaceFound } from "./traversal"
import type {
  BaseDimensions,
  MazeDimensions,
  CellAddress,
  CellNeighbors,
  Direction,
  LevelDimensions,
  NavigationProfile,
  PathStep,
  RenderGridPoint,
  RoundState,
  WallWeight,
} from "./types"

const { generation, maze: mazeConfig, scoring } = CONFIG

// getRandomNo returns a bounded, cryptographically random index for maze generation. Using
// crypto.getRandomValues rather than Math.random keeps maze layouts genuinely unpredictable —
// worthwhile even for a game, since a guessable layout would blunt the challenge, and it means
// no swap is needed later if the project grows into a context where that unpredictability
// becomes a real security property rather than just a gameplay one.
function getRandomNo(limit: number): number {
  if (limit <= 0) {
    return 0
  }

  // Rejection sampling avoids modulo bias: values landing in the partial final range above the
  // largest multiple of `limit` are discarded and re-rolled rather than folded in unevenly.
  const range = 2 ** 32
  const rejectionThreshold = range - (range % limit)
  const buffer = new Uint32Array(1)

  let value: number
  do {
    crypto.getRandomValues(buffer)
    value = buffer[0]
  } while (value >= rejectionThreshold)

  return value % limit
}

// absInt normalizes signed values when comparing candidate dimensions.
function absInt(value: number): number {
  return value < 0 ? -value : value
}

// getWallCharacters resolves the glyph set for the requested wall weight.
function getWallCharacters(weight: WallWeight): [string, string, string] {
  return mazeConfig.walls[weight]
}

// generateMazeArea turns a level number into the target maze area. Exported so the benchmark can
// find which level a case belongs to using this function rather than restating seed and diff.
export function generateMazeArea(level: number): number {
  return level * generation.diff + generation.seed
}

// appendFittingDimensions records factor pairs that still fit the viewport.
function appendFittingDimensions(
  candidates: BaseDimensions[],
  numCols: number,
  numRows: number,
  terminalSize: BaseDimensions,
): BaseDimensions[] {
  // Reject very skinny factor pairs even when they fit, because they make poor playable mazes.
  if (numCols < mazeConfig.minMazeSideCells || numRows < mazeConfig.minMazeSideCells) {
    return candidates
  }

  if (terminalSize.numCols >= numCols && terminalSize.numRows >= numRows) {
    candidates.push({ numCols, numRows })
  }

  const rotatedDimensions = { numCols: numRows, numRows: numCols }
  // Also try the same factor pair rotated, because a wide viewport may fit numRows x numCols better.
  if (
    numCols !== numRows &&
    terminalSize.numCols >= rotatedDimensions.numCols &&
    terminalSize.numRows >= rotatedDimensions.numRows
  ) {
    candidates.push(rotatedDimensions)
  }

  return candidates
}

// fittingDimensionsForArea enumerates all viewport-safe factor pairs for an area.
function fittingDimensionsForArea(
  area: number,
  terminalSize: BaseDimensions,
): BaseDimensions[] {
  const candidates: BaseDimensions[] = []

  for (
    let divisor = Math.floor(Math.sqrt(area));
    divisor >= mazeConfig.minMazeSideCells;
    divisor -= 1
  ) {
    if (area % divisor !== 0) {
      continue
    }

    appendFittingDimensions(candidates, divisor, Math.floor(area / divisor), terminalSize)
  }

  return candidates
}

// aspectMismatchScore measures how far a candidate is from the viewport aspect ratio.
function aspectMismatchScore(
  candidate: BaseDimensions,
  terminalSize: BaseDimensions,
): number {
  return absInt(
    candidate.numCols * terminalSize.numRows - candidate.numRows * terminalSize.numCols,
  )
}

// isPreferredMazeDimensions ranks one candidate against the current best fit.
function isPreferredMazeDimensions(
  candidate: BaseDimensions,
  currentBest: BaseDimensions,
  terminalSize: BaseDimensions,
): boolean {
  // Prefer the less skewed (close to square) shape so the maze stays balanced and playable.
  const candidateSkew = absInt(candidate.numCols - candidate.numRows)
  const bestSkew = absInt(currentBest.numCols - currentBest.numRows)
  if (candidateSkew !== bestSkew) {
    return candidateSkew < bestSkew
  }

  // When skew ties, choose the shape that better matches the viewport.
  const candidatePenalty = aspectMismatchScore(candidate, terminalSize)
  const bestPenalty = aspectMismatchScore(currentBest, terminalSize)
  if (candidatePenalty !== bestPenalty) {
    return candidatePenalty < bestPenalty
  }

  // Remaining ties are equivalent enough that the first deterministic candidate should stay selected.
  return false
}

// chooseBestMazeDimensions picks the most balanced candidate for the viewport.
function chooseBestMazeDimensions(
  candidates: BaseDimensions[],
  terminalSize: BaseDimensions,
): BaseDimensions {
  let best = candidates[0]

  for (const candidate of candidates.slice(1)) {
    if (isPreferredMazeDimensions(candidate, best, terminalSize)) {
      best = candidate
    }
  }

  return best
}

// resolveMazeArea keeps isolated bad factor pairs playable without borrowing the next level.
function resolveMazeArea(
  level: number,
  terminalSize: BaseDimensions,
): BaseDimensions[] | null {
  // Each visible level owns the area band from its raw target up to the next level's target.
  const area = generateMazeArea(level)
  const areaLimit = generateMazeArea(level + 1)
  const terminalArea = terminalSize.numRows * terminalSize.numCols
  if (area > terminalArea) {
    return null
  }

  const exactCandidates = fittingDimensionsForArea(area, terminalSize)
  if (exactCandidates.length > 0) {
    return exactCandidates
  }

  // We only get here after the current exact target missed.
  // If the next exact target also misses, stop early to preserve UI room for controls and seats.
  const nextExactCandidates = fittingDimensionsForArea(areaLimit, terminalSize)
  if (nextExactCandidates.length === 0) {
    return null
  }

  // Repair isolated bad factors inside this level's band without consuming the next level's area.
  for (
    let candidateArea = area + 1;
    candidateArea < areaLimit && candidateArea <= terminalArea;
    candidateArea += 1
  ) {
    const candidates = fittingDimensionsForArea(candidateArea, terminalSize)
    if (candidates.length > 0) {
      return candidates
    }
  }

  return null
}

// getMazeDimensions finds the best playable dimensions for a level and viewport.
export function getMazeDimensions(
  level: number,
  terminalSize: BaseDimensions,
): LevelDimensions | null {
  const candidates = resolveMazeArea(level, terminalSize)
  if (candidates === null) {
    return null
  }

  const selected = chooseBestMazeDimensions(candidates, terminalSize)
  return { ...createMazeDimensions(selected), level }
}

// createPlayingField builds the initial fully-walled maze grid.
function createPlayingField(
  dimensions: BaseDimensions,
  weight: WallWeight,
): string[][] {
  const chars = getWallCharacters(weight)
  const rows = mazeConfig.renderCellStep * dimensions.numRows + 1
  const horizontalOpening = " ".repeat(mazeConfig.wallOpening.horizontal)
  const data: string[][] = []

  for (let rowIndex = 0; rowIndex < rows; rowIndex += 1) {
    const row: string[] = []

    for (
      let columnIndex = 0;
      columnIndex <= dimensions.numCols;
      columnIndex += 1
    ) {
      row.push(chars[0])

      if (columnIndex !== dimensions.numCols && rowIndex % 2 === 0) {
        row.push(chars[1])
      } else if (columnIndex !== dimensions.numCols) {
        row.push(horizontalOpening)
      }
    }

    data.push(row)
  }

  return data
}

// getCellAddress maps a cell number to its wall and center coordinates.
function getCellAddress(
  dimensions: MazeDimensions,
  cellNo: number,
): CellAddress | null {
  if (cellNo <= 0 || cellNo > dimensions.area) {
    return null
  }

  const row = (Math.floor((cellNo - 1) / dimensions.numCols) + 1) * mazeConfig.renderCellStep
  const column = (((cellNo - 1) % dimensions.numCols) + 1) * mazeConfig.renderCellStep

  return {
    __topLeft: { x: column - mazeConfig.renderCellStep, y: row - mazeConfig.renderCellStep },
    __topCenter: { x: column - 1, y: row - mazeConfig.renderCellStep },
    __topRight: { x: column, y: row - mazeConfig.renderCellStep },
    __middleLeft: { x: column - mazeConfig.renderCellStep, y: row - 1 },
    __middleCenter: { x: column - 1, y: row - 1 },
    __middleRight: { x: column, y: row - 1 },
    __bottomLeft: { x: column - mazeConfig.renderCellStep, y: row },
    __bottomCenter: { x: column - 1, y: row },
    __bottomRight: { x: column, y: row },
  }
}

// getCellNeighbors returns the adjacent cell numbers around one cell.
function getCellNeighbors(
  dimensions: MazeDimensions,
  cellNo: number,
): CellNeighbors {
  if (cellNo <= 0 || cellNo > dimensions.area) {
    return { __bottom: 0, __left: 0, __right: 0, __top: 0 }
  }

  const column = (cellNo - 1) % dimensions.numCols
  const neighbors: CellNeighbors = {
    __bottom: 0,
    __left: 0,
    __right: 0,
    __top: 0,
  }

  if (column < dimensions.numCols - 1) {
    neighbors.__right = cellNo + 1
  }

  if (column > 0) {
    neighbors.__left = cellNo - 1
  }

  if (cellNo - dimensions.numCols > 0) {
    neighbors.__top = cellNo - dimensions.numCols
  }

  if (cellNo + dimensions.numCols <= dimensions.area) {
    neighbors.__bottom = cellNo + dimensions.numCols
  }

  return neighbors
}

// countNeighbors counts how many edges a cell exposes to the grid.
function countNeighbors(neighbors: CellNeighbors): number {
  let count = 0

  if (neighbors.__bottom !== 0) {
    count += 1
  }

  if (neighbors.__left !== 0) {
    count += 1
  }

  if (neighbors.__right !== 0) {
    count += 1
  }

  if (neighbors.__top !== 0) {
    count += 1
  }

  return count
}

// getPresentNeighbors filters neighboring cells down to the unvisited options.
function getPresentNeighbors(
  dimensions: MazeDimensions,
  cellNo: number,
  visited: boolean[],
): number[] {
  const neighbors = getCellNeighbors(dimensions, cellNo)
  const present: number[] = []

  if (neighbors.__bottom !== 0 && !visited[neighbors.__bottom]) {
    present.push(neighbors.__bottom)
  }

  if (neighbors.__left !== 0 && !visited[neighbors.__left]) {
    present.push(neighbors.__left)
  }

  if (neighbors.__right !== 0 && !visited[neighbors.__right]) {
    present.push(neighbors.__right)
  }

  if (neighbors.__top !== 0 && !visited[neighbors.__top]) {
    present.push(neighbors.__top)
  }

  return present
}

// countPresentNeighbors reports how many of a cell's neighbors are still unvisited. The
// least-neighbors bias calls this once per candidate on every generation step and only needs the
// count, so it avoids the intermediate array getPresentNeighbors would allocate.
function countPresentNeighbors(
  dimensions: MazeDimensions,
  cellNo: number,
  visited: boolean[],
): number {
  const neighbors = getCellNeighbors(dimensions, cellNo)
  let count = 0

  if (neighbors.__bottom !== 0 && !visited[neighbors.__bottom]) {
    count += 1
  }

  if (neighbors.__left !== 0 && !visited[neighbors.__left]) {
    count += 1
  }

  if (neighbors.__right !== 0 && !visited[neighbors.__right]) {
    count += 1
  }

  if (neighbors.__top !== 0 && !visited[neighbors.__top]) {
    count += 1
  }

  return count
}

// getNavigationProfile maps maze area into the same smooth difficulty curve used
// by the Go runtime. Smaller mazes keep longer corridors, while larger mazes
// tighten the limits until they reach the hardest supported profile.
export function getNavigationProfile(
  dimensions: MazeDimensions,
): NavigationProfile {
  const difficultyFactor = navigationDifficultyFactor(dimensions.area)

  return {
    __maxCorridorLength: interpolateNavigationValue(
      generation.navigation.friendlyProfile.__maxCorridorLength,
      generation.navigation.hardestProfile.__maxCorridorLength,
      difficultyFactor,
    ),
    __leastNeighborsBias: interpolateNavigationValue(
      generation.navigation.friendlyProfile.__leastNeighborsBias,
      generation.navigation.hardestProfile.__leastNeighborsBias,
      difficultyFactor,
    ),
  }
}

// navigationDifficultyFactor normalizes maze area into a 0..1 difficulty value.
function navigationDifficultyFactor(area: number): number {
  if (area <= generation.navigation.friendlyMaxArea) {
    return 0
  }

  if (area >= generation.navigation.hardestArea) {
    return 1
  }

  const normalizedArea = (area - generation.navigation.friendlyMaxArea) /
    (generation.navigation.hardestArea - generation.navigation.friendlyMaxArea)

  return Math.sqrt(normalizedArea)
}

// interpolateNavigationValue blends between the friendly and hardest profile values.
function interpolateNavigationValue(
  friendly: number,
  hardest: number,
  difficultyFactor: number,
): number {
  return Math.round(friendly + (hardest - friendly) * difficultyFactor)
}

// directionBetween converts two adjacent cells into a movement direction.
function directionBetween(
  dimensions: MazeDimensions,
  currentCell: number,
  nextCell: number,
): Direction {
  const neighbors = getCellNeighbors(dimensions, currentCell)

  switch (nextCell) {
    case neighbors.__top:
      return "MoveUp"
    case neighbors.__bottom:
      return "MoveDown"
    case neighbors.__left:
      return "MoveLeft"
    case neighbors.__right:
      return "MoveRight"
    default:
      return "none"
  }
}

// backtrackToBranch rewinds the carved path until an unvisited branch is found.
function backtrackToBranch(
  dimensions: MazeDimensions,
  path: PathStep[],
  visited: boolean[],
): { path: PathStep[]; currentCell: number; neighbors: number[] } {
  while (path.length > 0) {
    const currentCell = path[path.length - 1].__cellNo
    const neighbors = getPresentNeighbors(dimensions, currentCell, visited)

    if (neighbors.length > 0) {
      return { path, currentCell, neighbors }
    }

    path.pop()
  }

  throw new Error("maze generation failed: no branch with unvisited neighbors found")
}

// chooseNextCell applies the navigation profile to the next branch decision.
function chooseNextCell(
  dimensions: MazeDimensions,
  neighbors: number[],
  currentState: PathStep,
  profile: NavigationProfile,
  visited: boolean[],
): PathStep {
  const allChoices: PathStep[] = []
  const withinLengthLimit: PathStep[] = []

  for (const neighbor of neighbors) {
    const choice: PathStep = {
      __cellNo: neighbor,
      __moveDirection: directionBetween(dimensions, currentState.__cellNo, neighbor),
      __corridorLength: 1,
    }

    if (choice.__moveDirection === currentState.__moveDirection) {
      choice.__corridorLength += currentState.__corridorLength
    }

    allChoices.push(choice)

    // Caps how long a straight run can go before being forced to bend.
    if (choice.__corridorLength <= profile.__maxCorridorLength) {
      withinLengthLimit.push(choice)
    }
  }

  const choices = withinLengthLimit.length > 0 ? withinLengthLimit : allChoices
  if (choices.length > 1 && getRandomNo(scoring.percentScale) < profile.__leastNeighborsBias) {
    // Prefer the candidate with the fewest remaining unvisited neighbors of its own. A
    // low-neighbor-count cell gets "used up" cleanly by visiting it now, leaving nothing behind
    // for some later, unrelated branch to claim and retroactively turn this cell into a
    // junction. This is the mechanism that actually controls branching — unlike corridor length
    // or turn direction, neighbor count directly predicts whether a cell will be orphaned.
    let leastPopulated: PathStep[] = []
    let fewestRemaining = Infinity

    for (const choice of choices) {
      const remaining = countPresentNeighbors(dimensions, choice.__cellNo, visited)
      if (remaining < fewestRemaining) {
        fewestRemaining = remaining
        leastPopulated = [choice]
      } else if (remaining === fewestRemaining) {
        leastPopulated.push(choice)
      }
    }
    return leastPopulated[getRandomNo(leastPopulated.length)]
  }

  return choices[getRandomNo(choices.length)]
}

// getStartPosition prefers an edge cell so the opening feels less uniform.
function getStartPosition(dimensions: MazeDimensions): number {
  while (true) {
    const randomCellNo = getRandomNo(dimensions.area) + 1
    if (countNeighbors(getCellNeighbors(dimensions, randomCellNo)) < 4) {
      return randomCellNo
    }
  }
}

// createPath removes the wall segment between two connected cells.
function createPath(
  dimensions: MazeDimensions,
  maze: string[][],
  currentCellNo: number,
  nextCellNo: number,
): void {
  const address = getCellAddress(dimensions, currentCellNo)
  if (!address) {
    return
  }

  const horizontalOpening = " ".repeat(mazeConfig.wallOpening.horizontal)
  const verticalOpening = " ".repeat(mazeConfig.wallOpening.vertical)
  const neighbors = getCellNeighbors(dimensions, currentCellNo)

  switch (nextCellNo) {
    case neighbors.__bottom:
      maze[address.__bottomCenter.y][address.__bottomCenter.x] = horizontalOpening
      break
    case neighbors.__left:
      maze[address.__middleLeft.y][address.__middleLeft.x] = verticalOpening
      break
    case neighbors.__right:
      maze[address.__middleRight.y][address.__middleRight.x] = verticalOpening
      break
    case neighbors.__top:
      maze[address.__topCenter.y][address.__topCenter.x] = horizontalOpening
      break
  }
}

// replaceChar swaps a junction glyph only when the vertical path stays open.
function replaceChar(
  dimensions: MazeDimensions,
  point: RenderGridPoint,
  replacement: string,
  maze: string[][],
): void {
  let topItem = ""
  let bottomItem = ""
  let hasTop = false
  let hasBottom = false

  if (point.y - 1 > 0) {
    topItem = maze[point.y - 1][point.x]
    hasTop = true
  }

  if (point.y + 1 <= dimensions.numRows * mazeConfig.renderCellStep) {
    bottomItem = maze[point.y + 1][point.x]
    hasBottom = true
  }

  const row = point.y
  const column = point.x

  if (!hasTop && hasBottom && isSpaceFound(bottomItem)) {
    maze[row][column] = replacement
  } else if (hasTop && !hasBottom && isSpaceFound(topItem)) {
    maze[row][column] = replacement
  } else if (
    hasTop &&
    hasBottom &&
    isSpaceFound(topItem) &&
    isSpaceFound(bottomItem)
  ) {
    maze[row][column] = replacement
  }
}

// optimizeMaze softens eligible vertical joints after the maze is carved.
function optimizeMaze(
  dimensions: MazeDimensions,
  weight: WallWeight,
  maze: string[][],
): void {
  const chars = getWallCharacters(weight)

  for (let cell = 1; cell <= dimensions.area; cell += 1) {
    const address = getCellAddress(dimensions, cell)
    if (!address) {
      continue
    }

    replaceChar(dimensions, address.__bottomRight, chars[2], maze)
    replaceChar(dimensions, address.__topRight, chars[2], maze)
  }
}

// generateMaze carves the maze, then returns the grid plus start and target positions.
//
// The target is deliberately chosen as the single farthest cell from the start, by tree
// distance, across the entire maze — not an arbitrary or independently-random pick. Because a
// perfect maze is a spanning tree, and `path` here always mirrors the exact tree-path from
// `startCell` to whichever cell the DFS currently stands on, tracking `path.length`'s running
// maximum re-derives every cell's true distance from the start as it's first visited. The
// classic graph-theory guarantee for trees is that the farthest node from any fixed point is
// always an endpoint of the tree's longest possible path (its diameter) — so `startCell` and
// `finalCell` are always as far apart as the maze's shape allows, regardless of where the
// randomly-chosen `startCell` happens to land.
// profileOverride carves under a caller-supplied profile instead of the one getNavigationProfile
// derives from area. Gameplay always omits it; it exists for measurement, since the profile is a
// pure function of area and nothing else can hold the grid fixed while moving a knob — which is what
// separates a knob's effect from the grid's own.
export function generateMaze(
  dimensions: LevelDimensions,
  weight: WallWeight,
  profileOverride?: NavigationProfile,
): RoundState {
  const navigationProfile = profileOverride ?? getNavigationProfile(dimensions)
  const visited = new Array<boolean>(dimensions.area + 1).fill(false)
  const maze = createPlayingField(dimensions, weight)
  const startCell = getStartPosition(dimensions)

  let path: PathStep[] = [
    { __cellNo: startCell, __moveDirection: "none", __corridorLength: 0 },
  ]
  let currentCell = startCell
  let visitedCount = 1
  let longestPathLength = 1
  let finalCell = startCell

  const startAddress = getCellAddress(dimensions, startCell)
  if (!startAddress) {
    throw new Error("failed to resolve start address")
  }

  visited[currentCell] = true

  while (visitedCount < dimensions.area) {
    const backtrackedState = backtrackToBranch(dimensions, path, visited)
    path = backtrackedState.path
    currentCell = backtrackedState.currentCell

    const nextChoice = chooseNextCell(
      dimensions, backtrackedState.neighbors, path[path.length - 1], navigationProfile, visited,
    )

    if (visited[nextChoice.__cellNo]) {
      continue
    }

    visited[nextChoice.__cellNo] = true
    visitedCount += 1
    createPath(dimensions, maze, currentCell, nextChoice.__cellNo)
    path.push(nextChoice)

    if (path.length > longestPathLength) {
      longestPathLength = path.length
      finalCell = nextChoice.__cellNo
    }
  }

  const finalAddress = getCellAddress(dimensions, finalCell)
  if (!finalAddress) {
    throw new Error("failed to resolve target address")
  }

  optimizeMaze(dimensions, weight, maze)

  return {
    maze,
    startPosition: {
      x: startAddress.__middleCenter.x,
      y: startAddress.__middleCenter.y,
    },
    finalPosition: {
      x: finalAddress.__middleCenter.x,
      y: finalAddress.__middleCenter.y,
    },
  }
}
